### DISCUSSION TITLE
Iterative problem

### DISCUSSION THREAD

**Discussion by gabburgio** [idaholab/moose#24124] [Q&A General] [answered]
Hello everyone

I am trying to build an app that implements the procedure to calculate the SPH factors described in https://www.sciencedirect.com/science/article/pii/S0306454919300040 .
For the solution aspect, https://www.sciencedirect.com/science/article/pii/S0306454917303079 is also useful.

My question is, what MOOSE features are best for this purpose?

The procedure is based on iterations in which the multigroup neutron diffusion equation is solved, and then the obtained flux and current values are used to calculate corrective factors for material properties and boundary conditions, and possible source terms with which to solve it again, until convergence of the factors.

Is it possible to achieve this simply using actions, by having an action that adds problems until convergence?
Should I use "MultiApps" with a single subapplication?
Can a ReferenceResidualProblem work instead?

Thanks for your time

Gabriele

**Comment by lindsayad** [comment]
@vincentlaboure @YaqiWang @snschune do you guys want to weigh-in here?

**Comment by GiudGiud** [comment]
I would recommend setting this up either as an optimization problem with the optimization module https://mooseframework.inl.gov/modules/optimization/index.html
Or just do fixed point iterations and use a parametric study from the stochastic tools module
https://mooseframework.inl.gov/syntax/ParameterStudy/index.html

For very large SPH factor problems, there is large scale more complex machinery in Griffin to compute the SPH factors (I can find the paper if needed).
I have used successfully fixed point iterations to get SPH factors with OpenMOC. 
How many factors are you trying to determine?

**Comment by vincentlaboure** [comment, marked as answer]
The approach described in https://www.sciencedirect.com/science/article/pii/S0306454919300040 is relatively straightforward to perform in MOOSE as the equivalence (SPH, DF or SPH-DF) problem is recast into a steady-state source-free Newton formulation. In Griffin (previously, in Rattlesnake), we then just use a Steady executioner and solve for the fluxes (which the SPH or discontinuity factors are a by-product of).

For more complicated problems with tough convergence, it can be useful to first perform a few traditional SPH iterations (as described in Section 2.2.1 of https://www.sciencedirect.com/science/article/pii/S0306454917303079) to provide a good initial guess. Griffin has a dedicated solver to do that is fairly close to a Steady executioner.

**Comment by gabburgio** [comment]
Thanks for the pointers!
I am now implementing the SPH-PJFNK and I have a couple followup questions:

If I understood correctly, the SPH factors are supposed to be incorporated in the material properties. I tried to achieve this by passing an ElementIntegralArrayVariablePostprocessor to the material as a parameter and then using it while computing material properties:

```
#include "DummySPHMaterial.h"

registerMooseObject("pertApp",DummySPHMaterial);

InputParameters
DummySPHMaterial::validParams()
{
    InputParameters params = Material::validParams();
    
    params.addRequiredParam<RealEigenVector>("ref_diffusivity","Serpent 2 diffusion coefficient for each energy group in this material (used as starting value in the iteration)");
    params.addRequiredParam<RealEigenVector>("ref_sigma_r","Serpent 2 removal cross section for each energy group in this material");
    
    params.addRequiredParam<RealEigenVector>("ref_phi_mg","reference flux integrated over the material region");
    params.addRequiredParam<PostprocessorValue>("zone_integrator","ElementIntegralArrayVariablePostprocessor to calculate the flux integral over the material zone");   
    
    // params.set<MooseEnum>("constant_on") = "SUBDOMAIN";
    
    return params;
}

DummySPHMaterial::DummySPHMaterial(const InputParameters & parameters) :
    Material(parameters),
    _v_diffusivity(     getParam<RealEigenVector>("ref_diffusivity")),
    _v_sigma_r(         getParam<RealEigenVector>("ref_sigma_r")),
    _v_ref_phi_mg(  getParam<RealEigenVector>("ref_phi_mg")),
    _zone_integrator(    getPostprocessorValue("zone_integrator")),

    _diffusivity(       declareProperty<RealEigenVector>("Dummy_diffusivity")),
    _sigma_r(           declareProperty<RealEigenVector>("Dummy_sigma_r"))

{
}


void
DummySPHMaterial::computeQpProperties()
{


    for(int i=0; i< _v_diffusivity.size(); i++ ) 
    {
    _zone_integrator.setParameters("component",i); 
    _diffusivity[_qp][i]    = (_v_ref_phi_mg[i]/_zone_integrator)*_v_diffusivity[i];
    _sigma_r[_qp][i]           = (_v_ref_phi_mg[i]/_zone_integrator)*_v_sigma_r[i];
    }
}
```


This however doesn't compile because I am just passing the postprocessor value. 
I don't understand how to pass the postprocessor itself as the parameter, and whether that is feasible in the first place.

I also thought about somehow integrating the component explicitly while computing material properties, but I'm not sure how to achieve that.


I also have a somewhat theoretical question: is it correct to say that the reason the SPH-PJFNK is feasible in moose is that, during each step of the linear solve, the Krylov subspace that is used is actually constructed using powers of the Jacobian approximation obtained with the "latest iteration" of the residual matrix, i.e. a version in which material properties that depend on the solution have been reevaluated with the values of the solution at the previous krylov step?

**Reply by GiudGiud** [reply]
Hello

you can pass a PostprocessorName then use the getPostprocessorValueByName routine and store this in a reference (that part is really important)

Guillaume

**Reply by GiudGiud** [reply]
also `<RealEigenVector>` isnt what's usually used for vector. We use `std::vector<Real>`. 

Finally, are you writing the equations in terms of the SPH factors being the nonlinear variables?
Because otherwise I don't see how this is SPH-PJFNK

**Reply by gabburgio** [reply]
The approach (which is necessary unless I misunderstood how SPH-PJFNK is implemented in MOOSE) was to insert the SPH factors in the material property evaluation.
So for a given material, instead of it producing a cross section property Sigma, I wanted to have it produce a property mu*Sigma, where mu is the SPH factor, calculated as mu = (phi_ref)/(phi).

Clearly, to do so the material would have had to take as parameters the reference fluxes, the flux array variable itself, and possibly a postprocessor with which to perform the integrals (although maybe it would have worked better to just instantiate a postprocessor as an attribute of the material object).

To my understanding, this would have resulted in the material property being recalculated (i.e. the residual matrix being reassembled) at all linear/nonlinear steps while solving for the flux (hence the question about the matrix used for the Krylov spaces).
This is what I assumed @vincentlaboure meant when talking about a steady-state source free reformulation (in that all terms of the equations would be brought to the left hand side, instead of first building the source as in traditional SPH, and the problem itself would be just a (nonlinear, because of the dependence of the material properties on the flux) steady one).

This might all be nonsense to be frank, I am dealing with a lot of black boxes here.
For instance I don't know if it possible to have a postprocessor calculate an integral at every iteration of the solver.
The fact that properties are recalculated at every calculation step I got from the "material system" page on the MOOSE site: 
"The properties are produced on demand, thus the computed values are always up to date. For example, a property that relies on a solution variable (e.g., thermal conductivity as function of temperature) will be computed with the current temperature during the solve iterations, so the properties are tightly coupled."
The overall approach is what I gathered from the "Algorithm 2" box in https://www.sciencedirect.com/science/article/pii/S0306454917303079

I should mention that I only wanted to apply this to problems with vacuum boundary, so no normalization was required.
I also didn't insert all necessary physical properties in the example I provided, just diffusivity and removal cross-section for simplicity, but clearly the real thing should also have the other parameters appearing in the multigroup diffusion equation.

**Reply by GiudGiud** [reply]
> For instance I don't know if it possible to have a postprocessor calculate an integral at every iteration of the solver.

See the execute_on parameter of the postprocessor. It s set by default to end of the timestep for execution, but you can make it be on linear or nonlinear iterations

> To my understanding, this would have resulted in the material property being recalculated (i.e. the residual matrix being reassembled) at all linear/nonlinear steps

This is the default as well for PJFNK / Newton.

**Reply by vincentlaboure** [reply]
> This might all be nonsense to be frank, I am dealing with a lot of black boxes here.

I think your understanding is pretty good. The nonlinear variables are indeed the fluxes (which is equivalent to solving for the SPH factors but with the advantage of having a very similar formulation when we compute and use the SPH factors).

Regarding your implementation, I'm sure you can make it work that way but the trickiest part of the implementation is to make sure you can keep track of which reference flux goes to which part of the domain (at least if you are targeting large problems with lots of SPH regions). We use an extra element integer in the mesh to do that. Then you can have a single `ElementUserObject` performing the flux integration at each `linear` iteration over all the SPH regions.

**Reply by YaqiWang** [reply]
The postprocessors or the `ElementUserObject` performing the integration of fluxes on SPH regions can have `execute_on` parameter equal to `linear` or `nonlinear` as said by @GiudGiud. In the former, cross sections depends on SPH factors which depends on flux, thus we introduce nonlinearity in the residual evaluation, PJFNK is using Newton iteration to solve this nonlinear problem. In the later, Newton iteration is equivalent with SPH iteration, i.e. at each iteration or linear solve, SPH factors are fixed. If you do need to evaluate discontinuity factors on vacuum boundaries, you may want to consider using `DomainUserObject` so that you can put volumetric flux integration and surface partial current integration in one place. Wish you a success on this implementation.

**Reply by gabburgio** [reply]
Many thanks to all of you for your advice.
I have managed to put together what I think is a basic working version, about which I have a few more questions:

-For the postprocessors, I have been using  ` execute_on = 'LINEAR NONLINEAR'` because it seems to me to do what is described in the paper. Is this correct?
-I was not sure on how to go about implementing an initial source for the first iteration (like in the paper), so to avoid getting a trivial solution I decided to simply set an initial condition for the variable as a constant equal to the average flux over the whole geometry, like: 
```
[Variables]
[./flux]
    type = ArrayMooseVariable
    components = 2
    initial_condition = '1.5 1.5'
[../]
[]
```
Am I correct in assuming that this sets the all the FE coefficients for the first PJFNK iteration to 1.5?
Is there any more sensible way to do this, for instance using a finite element interpolation of a piecewise constant function that is equal to the average flux on the various homogenized zones (which for me coincide with material blocks)?
I looked at the function system but I haven't found much.

-I have been using RealEigenVector and RealEigenMatrix as types for the material properties because it is more convenient for manipulations, e.g.:


```
   _chi_nu_sigma_f[_qp]    = (1/_k_ref)*_v_chi.array()* (_v_nu_sigma_f.array()*sph_factors.array()).transpose();
  
for (int j =0;  j< _v_diffusivity.size(); j++)
  {
      _sigma_s[_qp].row(j)    = (_v_sigma_s.row(j).array())*sph_factors.array();
  }

```
Is there any reason not to do this?

-I can calculate the sph factors from the reference fluxes and the postprocessor values, but it would be nice if there was an ArrayMaterialPropertyIntegralPostprocessor like the one to integrate array variables.
If I want that I have to code it myself right?

**Reply by YaqiWang** [reply]
> For the postprocessors, I have been using  execute_on = 'LINEAR NONLINEAR'

Typically just having `linear` is ok. I assume that having an extra evaluation at the beginning of each Newton iteration does not change things.

> Am I correct in assuming that this sets the all the FE coefficients for the first PJFNK iteration to 1.5?

It does set the initial values to the flux variables. The meaning of FE coefficients depends on the shape function family. For LAGRANGE, yes, all coefficients will be the values at the nodes, thus equal to 1.5. You can implement your own InitialCondition to use the reference flux to initialize the flux in a more sensible way. Or you can start from the simple initialization with few SPH iterations.

> Is there any reason not to do this?

You can use gperftools (MOOSE has a documentation page on how to do it) to do some profiling if performance is a concern. But I doubt this can cause issue.

> If I want that I have to code it myself right?

Yes, if it is missing in MOOSE. But if you think that can be generally useful and can be maintained in framework, please feel free to contribute by sending in a pull request to MOOSE repository on github.

**Comment by gabburgio** [comment]
I now have an implementation of PJFNK-SPH-DF that works for very simple problems, but doesn't seem to converge for a more complicated one, and I would like to try to add the traditional sph iterations at the beginning.
Is this best accomplished by using stateful material properties on the source terms?

I am now using ArrayConstantICs to initialize the flux to its average value on the homogenized zone. Is this likely to be too coarse?
Should the vacuum BC term be part of the source in the picard iteration?

Thanks again

Gabriele
